library/std/src/io/mod.rs

```diff
@@ -462,9 +462,14 @@ pub(crate) fn default_read_to_end<R: Read + ?Sized>(
             if read == 0 {
                 return Ok(buf.len() - start_len);
             }
+            // In the case of very short reads, continue to use the stack buffer
+            // until either we reach the end or we need to reallocate.
+            continue;
         }
 
-        if buf.len() == buf.capacity() {
+        // Try to ensure we're not making unnecessarily short reads.
+        // This is skipped if the user's pre-allocated Vec may have been exactly sized.
+        if buf.capacity() != start_cap || buf.spare_capacity_mut().len() == 0 {
```

Comment:
How does this new condition avoids the issue that's supposed to be fixed? With this code we can still do short reads. I still think that calling `try_reserve` without check for the spare capacity is the simplest, and keeping the check about `start_cap` avoids over-allocations.

Reply:
Hm, the only case where the spare capacity will be less than `PROBE_SIZE` after this point is when `small_probe_read` reallocates but the reallocation ends up smaller than `PROBE_SIZE`. Given `Vec`'s reallocation strategy. I'm not sure if that's an issue in practice given the loop above, I guess the `if` statement is indeed made redundant.